user-console/source/org/pentaho/mantle/client/commands/AdhocRunInBackgroundCommand.java

```diff
@@ -0,0 +1,233 @@
+    this.outputName = outputName;
+  }
+
+  @Override
+  protected void performOperation() {
+    final SolutionBrowserPanel sbp = SolutionBrowserPanel.getInstance();
+    if ( this.getSolutionPath() != null ) {
+      sbp.getFile( this.getSolutionPath(), new SolutionFileHandler() {
+        @Override
+        public void handle( RepositoryFile repositoryFile ) {
+          showDialog( true );
+        }
+      } );
+    } else {
+      performOperation( true );
+    }
+  }
+
+  @Override
+  protected void showDialog( final boolean feedback ) {
+    final ScheduleOutputLocationDialog outputLocationDialog = new ScheduleOutputLocationDialog( getSolutionPath() ) {
+      @Override
+      protected void onSelect( final String name, final String outputLocationPath ) {
+        setOutputName( name );
+        setOutputLocationPath( outputLocationPath );
+        performOperation( feedback );
+      }
+
+      @Override protected void onCancel() {
+        super.onCancel();
+        AdhocRunInBackgroundCommand.onCancel();
+      }
+
+      @Override protected void onOk() {
+        super.onOk();
+        AdhocRunInBackgroundCommand.onOk( getOutputLocationPath() );
+      }
+
+      @Override protected void onAttach() {
+        super.onAttach();
+        AdhocRunInBackgroundCommand.onAttach();
+      }
+    };
+
+    outputLocationDialog.setOkButtonText( Messages.getString( "ok" ) );
+    outputLocationDialog.center();
+  }
+
+  public static native void onCancel()
+  /*-{
+    $wnd.mantle_fireEvent('GenericEvent', {eventSubType: 'locationPromptCanceled'});
+  }-*/;
+
+  public static native void onOk( final String outputPath )
+  /*-{
+    $wnd.mantle_fireEvent('GenericEvent', {eventSubType: 'locationPromptOk', stringParam: outputPath});
+  }-*/;
+
+  public static native void onAttach()
+  /*-{
+    $wnd.mantle_fireEvent('GenericEvent', {eventSubType: 'locationPromptAttached'});
+  }-*/;
+
+
+  private RequestBuilder createTreeRequest() {
+    RequestBuilder scheduleFileRequestBuilder = new RequestBuilder( RequestBuilder.GET, contextURL + "api/repo/files/"
+      + NameUtils.encodeRepositoryPath( getOutputLocationPath() ) + "/tree?depth=1" );
+    scheduleFileRequestBuilder.setHeader( "If-Modified-Since", "01 Jan 1970 00:00:00 GMT" );
+    return scheduleFileRequestBuilder;
+  }
+
+  @Override
+  protected void performOperation( boolean feedback ) {
+
+    RequestBuilder treeRequestBuilder = createTreeRequest();
+
+    try {
+      treeRequestBuilder.sendRequest( null, new RequestCallback() {
+
+        public void onError( Request request, Throwable exception ) {
+          MessageDialogBox dialogBox =
+            new MessageDialogBox( Messages.getString( "error" ), exception.toString(), false, false,
+              true );
+          dialogBox.center();
+        }
+
+        public void onResponseReceived( Request request, Response response ) {
+          if ( response.getStatusCode() == Response.SC_OK ) {
+            String folderId = null;
```

Comment:
Could we extract this block into a "parseFolderId(Response response)" method? Makes the block a lot easier to read. Along with moving the error handling into extra methods (see above), this then becomes a trivially simple to read method. I love trivially simple - it matches my brain :)